src/fold.jl

```diff
@@ -0,0 +1,320 @@
+current distribution by returning `2c - x`
+- if `x` is outside the domain of the folded distribution, but inside the parent distribution, then this
+just returns `x`
+- if `x` is not in the support of the parent distribution, then the function gives a DomainError 
+"""
+function unfold_value(x::Real, d::Folded)
+    original_support_error_check(x,d) # Will error if x ∉ domain(d.original)
+    return insupport(d.included, x) ? (2*d.crease - x) : x
+end
+
+
+
+######################################################
+#         Distribution.jl extended methods
+######################################################
+"""
+    insupport(d::Folded, x::Real)
+
+Returns true if the value `x` is in the domain (or support) of the folded
+distribution.
+"""
+insupport(d::Folded, x::Real) = insupport(d.included, x)
+
+
+"""
+    minimum(d::Folded)
+
+Returns the lowest value of the domain of the distribution
+"""
+minimum(d::Folded) = minimum(d.included)
+
+
+
+"""
+    maximum(d::Folded)
+
+Returns the highest value of the domain of the distribution
+"""
+maximum(d::Folded) = maximum(d.included)
+
+
+"""
+    rand(::AbstractRNG, d::Folded)
+
+Returns samples from the folded distribution
+"""
+rand(rng::AbstractRNG, d::Folded) = fold_value(rand(rng,d.original), d)
+
+"""
+   logpdf(d::Folded, x::Real)
+
+Returns the logarithm of the pdf of the folded distribution. 
+Returns -Inf if the value is out of the distribution.
+
+Example:
+```julia
+logpdf(folded(Normal(μ,σ),0), -3) # -> -Inf
+logpdf(folded(Normal(1.0,1.0),0), 3) # -> 0.05412479673895295
+``` 
+"""
+function logpdf(d::Folded, x::Real)
+    !insupport(d,x) && return -Inf
+    return logaddexp(logpdf(d.original,x),logpdf(d.original,unfold_value(x,d)))
+end
+
+
+"""
+   pdf(d::Folded, x::Real)
+
+Returns the the pdf of the folded distribution. This is simply the
+sum of the value of the pdf of the original distribution at two locations:
+one at `x` and the other at the reflected point `2c-x`, where `c` is the crease.
+
+Example:
+```julia
+pdf(folded(Normal(μ,σ),0), -3) # -> 0.0
+pdf(folded(Normal(1.0,1.0),0), 3) # -> 0.05412479673895295
+pdf(folded(Normal(3,4),0.0),0.1) ≈ pdf(Normal(3,4),-0.1) + pdf(Normal(3,4),0.1) # -> true
+```
+"""
+function pdf(d::Folded, x::Real)
+    !insupport(d,x) && return zero(x)
+    #return pdf(d.original,x) + pdf(d.original,unfold_value(x,d)) -- removed for numerical stability
+    return exp(logpdf(d,x))
+end
+
+"""
+    cdf(d::Folded, x::Real) 
+
+Example:
+cdf(folded(Beta(α,β),0.5), -3) # -> 0.0
+cdf(folded(Beta(α,β),0.5), 3) # -> 1.0
+cdf(folded(Normal(1.0,1.0),0), 1.0) # -> 0.4772498680518208
+```
+"""
+function cdf(d::Folded, x::Real) 
+    if !insupport(d,x)
+        (x ≤ minimum(d)) && return zero(x) # Is below the support
+        (x ≥ maximum(d)) && return one(x) # Is above the support
+    end
+    d.included.tp*cdf(d.included, x) + d.excluded.tp*ccdf(d.excluded, unfold_value(x,d))
```

Comment:
d.included.tp does not exist for general truncated distributions. Could we use something like
```suggestion
cdf(d::Folded, x::Real) = exp(logcdf(d, x)) # not sure, maybe the default definition already?
function logcdf(d::Folded, x::Real)
    _x = clamp(x, extrema(d.included)...)
    y = unfolded_value(d, _x)
    a, b = d.keep_right ? (y, _x) : (_x, y)
    return logdiffcdf(d.original, b, a)
end
```

Reply:
My understanding, by looking at the source code was that the Truncated object has a field `tp` in its definition, and having this would prevent recalculation each time. 

Is there a way to infer from the types whether a truncated distribution has the `tp` field? So if it doesn't exist for a particular type, we could dispatch to the code you provided? 

https://github.com/JuliaStats/Distributions.jl/blob/fa8c30dedf058cb9891ef141c14cc857576221ac/src/truncate.jl#L93